service/src/parentchain_block_syncer.rs

```diff
@@ -0,0 +1,109 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+	Copyright (C) 2017-2019 Baidu, Inc. All Rights Reserved.
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+use crate::enclave_account;
+use itp_api_client_extensions::{AccountApi, ChainApi};
+use itp_enclave_api::{enclave_base::EnclaveBase, sidechain::Sidechain};
+use itp_types::SignedBlock;
+use log::{error, trace};
+use my_node_runtime::{BlockNumber, Header};
+use sp_core::sr25519;
+use substrate_api_client::{rpc::WsRpcClient, Api};
+
+const BLOCK_SYNC_BATCH_SIZE: u32 = 1000;
+
+pub(crate) struct ParentchainBlockSyncer {
+	api: Api<sr25519::Pair, WsRpcClient>,
+}
+
+impl ParentchainBlockSyncer {
+	pub fn new(api: Api<sr25519::Pair, WsRpcClient>) -> Self {
+		ParentchainBlockSyncer { api }
+	}
+
+	/// Fetches the amount of blocks to sync from the parentchain and feeds them to the enclave.
+	/// Returns the latest synced block Header.
+	///
+	pub fn sync_parentchain<E: EnclaveBase + Sidechain>(
+		&self,
+		enclave_api: &E,
```

Comment:
In a next step, the `enclave_api` can also be a member, not passed into this function. This will require you to introduce a type parameter into you struct: `EnclaveApi` with a new member `enclave_api: EnclaveApi`. And then use `where` to specify the trait bounds: `EnclaveBase + Sidechain`.